pytest_customization/plugins/ocscilib.py

```diff
@@ -0,0 +1,141 @@
+"""
```

Comment:
Would it make sense to make plugins directory placed directly in root of the repo?

Reply:
I mean, which other pytest customization could we have here other than plugins?

Reply:
I was thinking about defining for example markers module here as well and define all markers like

```python
import pytest

tier1 = pytest.mark.tier1
tier2 = pytest.mark.tier3
```

then from test modules we can import

```python
from pytest_customization.markers import tier1


@tier1
def test_tier1():
    pass
```

Cannot thing now about any other, cause hooks are like plugins but we can separate them to hooks package maybe.

Reply:
I would rather not do that. This is additional level of abstraction, which would increase complexity a bit without any benefit. When I look at `@tier1` function annotation, I have no idea what it actually means. But looking at `@pytest.mark.tier1` gives a clear indication that this is a pytest marker.

That said, maintaining list of valid markers is not a bad idea. We can do that via [Registering markers](https://docs.pytest.org/en/latest/example/markers.html#registering-markers), which can be combined with `--strict-markers` to prevent using unregistered markers (it could be enabled by default in pytest.ini as well).

Reply:
I think that  in latest pytest there is warning message automatically printed that you are using non registered markers, not sure about if it's just warning or it's by default strict. I am already registering those markers in pytest.ini file as you can see @mbukatov .

And about abstraction, we used it in RHV a lot and found it quite handy, cause it allowed us to define something like this:
```python
BEFORE_UPGRADE = 1
UPGRADE = 2
BEFORE_UPGRADE_HOSTS = 3
UPGRADE_HOSTS = 4
AFTER_UPGRADE_HOSTS = 5
UPGRADE_CL = 6
AFTER_UPGRADE_CL = 7
UPGRADE_DC = 8
AFTER_UPGRADE = 9

# Tests ordering
FIRST_TEST = AFTER_UPGRADE + 1
TEST = FIRST_TEST + 1
LAST_TEST = TEST + 1

# Polarion decorator
polarion = pytest.mark.polarion

# Bugzilla decorator
# https://github.com/rhevm-qe-automation/pytest_marker_bugzilla
bz = pytest.mark.bugzilla

# Jira decorator
# https://github.com/vkondula/pytest_jira/
jira = pytest.mark.jira

network = pytest.mark.network
sla = pytest.mark.sla
storage = pytest.mark.storage
coresystem = pytest.mark.coresystem
virt = pytest.mark.virt
integration = pytest.mark.integration

# used for tests that are not adjusted to GE or tests that we don't want to run
do_not_run = pytest.mark.do_not_run(value=17)
endless = pytest.mark.endless(value='endless')
upgrade = pytest.mark.upgrade(value='upgrade')
tier1 = pytest.mark.tier1(value=1)
tier2 = pytest.mark.tier2(value=2)
tier3 = pytest.mark.tier3(value=3)
tier4 = pytest.mark.tier4(value=4)
timeout = pytest.mark.timeout
storages = pytest.mark.storages

tier_markers = [tier1, tier2, tier3, tier4, upgrade, do_not_run]
team_markers = [network, sla, storage, coresystem, virt, upgrade]

# order markers for ordering tests
order_before_upgrade = pytest.mark.run(order=BEFORE_UPGRADE)
order_upgrade = pytest.mark.run(order=UPGRADE)
order_before_upgrade_hosts = pytest.mark.run(order=BEFORE_UPGRADE_HOSTS)
order_upgrade_hosts = pytest.mark.run(order=UPGRADE_HOSTS)
order_after_upgrade_hosts = pytest.mark.run(order=AFTER_UPGRADE_HOSTS)
order_upgrade_cluster = pytest.mark.run(order=UPGRADE_CL)
order_after_upgrade_cluster = pytest.mark.run(order=AFTER_UPGRADE_CL)
```

We had some plugins which for example added default timeout for different tiers.
Or thanks to value=X we were able to easily know which number of test is directly from test.
Another usage was about ordering tests for pytest order plugin as you can see our order markers for upgrade tests.

So instead of ppl decorated the tests all the time like @pytest.mark.run(order=UPGRADE_CL) they just used @order_upgrade_cluster .

So this will is advantage we used for the markers defined like this and I think that ppl will know about those markers if we will document them properly.

@RazTamir  WDYT here?

Reply:
I'm used to this way as this is how we worked in RHV so +1.
If there were any issues with that in the past, I don't know but if you know @petr-balogh so maybe we can enhance it here as part of the implementation.

In overall, using @tier1 is very easy usage imo and people will get used to it (@mbukatov comment above).

This whole set provides a great flexibility

Reply:
I see your point with odering markers, but otherwise I'm not convinced that it's a good idea.

I'm curious what is this about: `pytest.mark.tier1(value=1)`. What is this `value=1` used for, when the marker is called `tier1` already? I know that one can read this value from fixture or test code, but what is the point here exactly?

Reply:
> I think that in latest pytest there is warning message automatically printed that you are using non registered markers, not sure about if it's just warning or it's by default strict. I am already registering those markers in pytest.ini file as you can see @mbukatov .
> 
> And about abstraction, we used it in RHV a lot and found it quite handy, cause it allowed us to define something like this:
> 
> ```python
> BEFORE_UPGRADE = 1
> UPGRADE = 2
> BEFORE_UPGRADE_HOSTS = 3
> UPGRADE_HOSTS = 4
> AFTER_UPGRADE_HOSTS = 5
> UPGRADE_CL = 6
> AFTER_UPGRADE_CL = 7
> UPGRADE_DC = 8
> AFTER_UPGRADE = 9
> 
> # Tests ordering
> FIRST_TEST = AFTER_UPGRADE + 1
> TEST = FIRST_TEST + 1
> LAST_TEST = TEST + 1
> 
> # Polarion decorator
> polarion = pytest.mark.polarion
> 
> # Bugzilla decorator
> # https://github.com/rhevm-qe-automation/pytest_marker_bugzilla
> bz = pytest.mark.bugzilla
> 
> # Jira decorator
> # https://github.com/vkondula/pytest_jira/
> jira = pytest.mark.jira
> 
> network = pytest.mark.network
> sla = pytest.mark.sla
> storage = pytest.mark.storage
> coresystem = pytest.mark.coresystem
> virt = pytest.mark.virt
> integration = pytest.mark.integration
> 
> # used for tests that are not adjusted to GE or tests that we don't want to run
> do_not_run = pytest.mark.do_not_run(value=17)
> endless = pytest.mark.endless(value='endless')
> upgrade = pytest.mark.upgrade(value='upgrade')
> tier1 = pytest.mark.tier1(value=1)
> tier2 = pytest.mark.tier2(value=2)
> tier3 = pytest.mark.tier3(value=3)
> tier4 = pytest.mark.tier4(value=4)
> timeout = pytest.mark.timeout
> storages = pytest.mark.storages
> 
> tier_markers = [tier1, tier2, tier3, tier4, upgrade, do_not_run]
> team_markers = [network, sla, storage, coresystem, virt, upgrade]
> 
> # order markers for ordering tests
> order_before_upgrade = pytest.mark.run(order=BEFORE_UPGRADE)
> order_upgrade = pytest.mark.run(order=UPGRADE)
> order_before_upgrade_hosts = pytest.mark.run(order=BEFORE_UPGRADE_HOSTS)
> order_upgrade_hosts = pytest.mark.run(order=UPGRADE_HOSTS)
> order_after_upgrade_hosts = pytest.mark.run(order=AFTER_UPGRADE_HOSTS)
> order_upgrade_cluster = pytest.mark.run(order=UPGRADE_CL)
> order_after_upgrade_cluster = pytest.mark.run(order=AFTER_UPGRADE_CL)
> ```
> 
> We had some plugins which for example added default timeout for different tiers.
> Or thanks to value=X we were able to easily know which number of test is directly from test.
> Another usage was about ordering tests for pytest order plugin as you can see our order markers for upgrade tests.
> 
> So instead of ppl decorated the tests all the time like @pytest.mark.run(order=UPGRADE_CL) they just used @order_upgrade_cluster .
> 
> So this will is advantage we used for the markers defined like this and I think that ppl will know about those markers if we will document them properly.
> 
> @RazTamir WDYT here?

Looking at the above ordering makes me *nervous*, Just to set the ordering right, is this the only way to do it? We are not only going to confuse the new comers here but unnecessarily create a complexity and for this reason alone its a  -1 from me.  I cannot believe how our previously simplicity of just ordering the tests in yaml solved this problem so neatly.

Reply:
> Looking at the above ordering makes me _nervous_, Just to set the ordering right, is this the only way to do it? We are not only going to confuse the new comers here but unnecessarily create a complexity and for this reason alone its a -1 from me. I cannot believe how our previously simplicity of just ordering the tests in yaml solved this problem so neatly.

Ordering test cases using markers is not a bad idea. That said, in our case with OCS, most of the ordering could be resolved via fixture tracking automatically and so explicit ordering via markers may not be that important anyway.

Reply:
@vasukulkarni The thing is that the tests in pytest way shouldn't be ordered at all and shouldn't depend on some previous one and should be triggered independently.  By default they will run in the order how they are defined in module class I think.

I am not telling that we should start using ordering now, I am just telling what is the advantage of have defined those markers on one place and just import them and mart test with @tier1.

In upgrade tests we needed to really have test ordered to do some operations before datacenter updated, before cluster updated and so on.. So we used this way how to order tests by order-plugin and just mark the tests in which order they should run in upgrade testing.

@mbukatov value=1 fro tier 1 was used in some of our pytest plugin we had and needed to consider the number of tier. So was easier to take it from value from pre defined marker which everyone used than to pare the number of tier from the name of marker. Also we had for example upgrade tear, and as you can see it doesn't contain any number in the marker.

At some point we will get to situation when we will need to inject some values to markers and I think that have them defined on one place will be easy to use them.

Reply:
@vasukulkarni I'll try to explain from a tester pov:

We can create a file with classes like:

```pyhon
@manage
class Manage()
    pass

@E2E
class e2e()
    pass
```

And in each test we will have something like:
```python
from testm_classes import Manage

@tier1  # This should be very easy to understand based on the test content or Polarin
class TestCaseOCS12345(Manage):  # this test will inherit the @manage decorator
```

The ordering can be done in the teams_classes.py like the team decorators.

So to recap, the only thing that a tester should take care of is adding a decorator on its class

Reply:
> @mbukatov value=1 fro tier 1 was used in some of our pytest plugin we had and needed to consider the number of tier. So was easier to take it from value from pre defined marker which everyone used than to pare the number of tier from the name of marker. Also we had for example upgrade tear, and as you can see it doesn't contain any number in the marker.
>
> At some point we will get to situation when we will need to inject some values to markers and I think that have them defined on one place will be easy to use them.

If you know beforehand that you need to process value of fixtures in plugins, it's easier to just use markers like `@pytest.marker.tier(value=1)`.

Reply:
@mbukatov yep, it's possible but we faced the issue with inheritance and reading the real value of marker kwargs value as this wasn't easily possible that time. But this issue was already fixed in pytetst I think.

Just found some related issues with markers we faced in past:
https://github.com/pytest-dev/pytest/issues/2515
https://github.com/pytest-dev/pytest/issues/3605

Reply:
@mbukatov another thing which I've just realized and which was the main point is that you cannot call pytest -m tier(1)   so that's why pytest -m tier1 and team for example

Reply:
@petr-balogh good point

Reply:
I added some structure in next commit. So please review once more @red-hat-storage/top-level-reviewers